src/menu/MenuItem.ts

```diff
@@ -0,0 +1,186 @@
+ * @property properties     Additional properties for the widget's vnode.
+ * @property selected       Determines whether the widget is marked as selected.
+ * @property tag            The HTML tag name to use for the widget's vnode. Defaults to 'a'.
+ * @property type           The type of the menu item. Defaults to 'item'.
+ */
+export interface MenuItemProperties extends ThemeableProperties {
+	active?: boolean;
+	controls?: string;
+	disabled?: boolean;
+	expanded?: boolean;
+	focusable?: boolean;
+	hasMenu?: boolean;
+	hasPopup?: boolean;
+	id?: string;
+	index?: number;
+	onClick?: (event: Event, index?: number) => void;
+	onKeyDown?: (event: KeyboardEvent, index?: number) => void;
+	onMouseDown?: (event: Event, index?: number) => void;
+	properties?: VirtualDomProperties;
+	selected?: boolean;
+	tag?: string;
+	type?: MenuItemType;
+}
+
+/**
+ * @private
+ * The maximum number of frames that should execute before a menu item scheduled to receive focus
+ * stops waiting to become visible.
+ */
+const MAX_FOCUS_FRAME_COUNT = 30;
+
+const roleMap: { [key: string]: string } = {
+	item: 'menuitem',
+	checkbox: 'menuitemcheckbox',
+	radio: 'menuitemradio'
+};
+
+export const MenuItemBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export class MenuItem extends MenuItemBase<MenuItemProperties> {
+	render() {
+		const {
+			controls,
+			disabled = false,
+			expanded = false,
+			focusable,
+			hasPopup = false,
+			hasMenu = false,
+			id,
+			index,
+			properties,
+			selected = false,
+			tag = 'a',
+			type = 'item'
+		} = this.properties;
+
+		const classes = this.classes(
+			hasMenu ? css.menuLabel : css.menuItem,
+			disabled ? css.disabled : null,
+			selected ? css.selected : null
+		);
+
+		const itemProperties: { [key: string]: any; } = {
+			'aria-controls': controls,
+			'aria-expanded': hasPopup ? String(expanded) : undefined,
+			'aria-haspopup': hasPopup ? 'true' : undefined,
+			'aria-disabled': disabled ? 'true' : undefined,
+			'data-dojo-index': typeof index === 'number' ? String(index) : undefined,
+			classes,
+			id,
+			key: 'root',
+			onclick: this._onClick,
+			onkeydown: this._onKeyDown,
+			onmousedown: this._onMouseDown,
+			role: roleMap[type],
+			tabIndex: focusable ? 0 : -1
+		};
+
+		if (type === 'checkbox' || type === 'radio') {
+			itemProperties['aria-checked'] = String(selected);
+		}
+
+		return v(tag, assign(Object.create(null), properties, itemProperties), this.children);
+	}
+
+	protected onElementUpdated(element: HTMLElement, key: string) {
+		if (key === 'root') {
+			let visibility = getComputedStyle(element).getPropertyValue('visibility');
+
+			if (this.properties.active) {
+				if (visibility === 'hidden') {
+					// When animating from `visibility: hidden`, some browsers will mark the element as visible
+					// within the next task, whereas others may take multiple tasks before the element switches
+					// to visible.
+					let callCount = 0;
+					const scheduleFocus = () => {
```

Comment:
Since rAF may need to be called multiple times before the element becomes visible in certain browsers, depending on the animation.